In [80]:
import numpy as np
import pandas as pd
import matplotlib as plt
import matplotlib.pyplot as plt
from matplotlib import pyplot
from numpy import std
import seaborn as sns
from scipy import stats

In [166]:
df = pd.read_csv('yelp_academic_dataset_reviews.csv')
df

review_id                 user_id  \
0       BiTunyQ73aT9WBnpR9DZGw  OyoGAe7OKpv6SyGZT5g77Q   
1       AqPFMleE6RsU23_auESxiA  _7bHUi9Uuf5__HHc_Q8guQ   
2       JrIxlS1TzJ-iCu79ul40cQ  eUta8W_HdHMXPzLBBZhL1A   
3       8JFGBuHMoiNDyfcxuWNtrA  smOvOajNG0lS4Pq7d8g4JQ   
4       oyaMhzBSwfGgemSGuZCdwQ  Dd1jQj7S-BFGqRbApFzCFw   
...                        ...                     ...   
967547  tWCi4N2qCil876BI2NDAZg  5d32aJe0a3_fSl3LIVYkWQ   
967548  nLjbVsETpqO17RbFcqskkA  am7-gkH_PDz598oTdYSD6A   
967549  99ylx-qPUSseITqBye2MpA  -AkziDwQ8hv2COTDBBUpig   
967550  KlHxcAifUF5zDKpJCBrRsw  7ziWZULyiZv2TesYNMFf4g   
967551  cACxcUY_AIsQKkpDRXuqnw  MCzlzlOw7IGbRAKVjJBPtg   

                   business_id  stars  useful  funny  cool  \
0       7ATYjTIgM3jUlt4UM3IypQ      5       1      0     1   
1       kxX2SOes4o-D3ZQBkiMRfA      5       1      0     1   
2       04UD14gamNjLY0IDYVhHJg      1       1      2     1   
3       RZtGWDLCAtuipwaZ-UfjmQ      4       0      0     0   
4       YtSqYv1Q_pOltsVPSx54SA      5       0      0     0   
...                        ...    ...     ...    ...   ...   
967547  EmrOCT7fArKI0JI_XF6aaw      1       5      2     0   
967548  3gVSrS4kffGGZT8oXHsIcw      3       2      0     2   
967549  aunmz06iWvo3bd6MMHEbqg      3       0      0     0   
967550  qQO7ErS_RAN4Vs1uX0L55Q      4       1      0     0   
967551  fcGexL5VH5G2Xw0tRj9uOQ      3       1      1     0   

                                                     text                 date  
0       I've taken a lot of spin classes over the year...  2012-01-03 15:28:18  
1       Wow!  Yummy, different,  delicious.   Our favo...  2015-01-04 00:01:03  
2       I am a long term frequent customer of this est...  2015-09-23 23:10:31  
3       Good food--loved the gnocchi with marinara\nth...  2009-10-14 19:57:14  
4       Tremendous service (Big shout out to Douglas) ...  2013-06-24 11:21:25  
...                                                   ...                  ...  
967547  Temple University blatantly disregards freedom...  2021-08-28 18:46:05  
967548  *Later Yelp* I've only been here once, but I l...  2014-11-03 14:45:46  
967549  Philly has become a dangerous place with the m...  2022-01-18 03:48:44  
967550  ice cream! ice cream sodas, sundaes!! \n\nwant...  2012-10-21 04:08:40  
967551  This is a good pizza option - they deliver thr...  2018-03-13 13:54:48  

[967552 rows x 9 columns]

In [167]:
df1 = pd.read_csv('yelp_academic_dataset_business.csv')
df1

business_id                                 name  \
0      MTSW4McQd7CbVtyjqoe9mw                   St Honore Pastries   
1      MUTTqe8uqyMdBl186RmNeA                             Tuna Bar   
2      ROeacJQwBeh05Rqg7F6TCg                                  BAP   
3      QdN72BWoyFypdGJhhI5r7g                              Bar One   
4      Mjboz24M9NlBeiOJKLEd_Q                     DeSandro on Main   
...                       ...                                  ...   
14564  wPAaAbFbq5wfLfW0gnYU7g                             Intermix   
14565  JDM7kiCEmeBc_5W-z6eZZw              2601 Parkway Apartments   
14566  9U1Igcpe954LoWZRmNc-zg  Hand & Stone Massage And Facial Spa   
14567  LJ4GjQ1HL6kqvIPpNUNNaQ             Shanti Yoga and Ayurveda   
14568  WnT9NIzQgLlILjPT0kEcsQ        Adelita Taqueria & Restaurant   

                            address          city state  postal_code  \
0                       935 Race St  Philadelphia    PA      19107.0   
1                       205 Race St  Philadelphia    PA      19106.0   
2                     1224 South St  Philadelphia    PA      19147.0   
3                      767 S 9th St  Philadelphia    PA      19147.0   
4                      4105 Main St  Philadelphia    PA      19127.0   
...                             ...           ...   ...          ...   
14564                1718 Walnut St  Philadelphia    PA      19103.0   
14565         2601 Pennsylvania Ave  Philadelphia    PA      19130.0   
14566  1100 S Columbus Blvd, Ste 24  Philadelphia    PA      19147.0   
14567            1638 Pine St, Fl 1  Philadelphia    PA      19103.0   
14568                 1108 S 9th St  Philadelphia    PA      19147.0   

        latitude  longitude  stars  review_count  is_open  \
0      39.955505 -75.155564    4.0            80        1   
1      39.953949 -75.143226    4.0           245        1   
2      39.943223 -75.162568    4.5           205        1   
3      39.939825 -75.157447    4.0            65        0   
4      40.022466 -75.218314    3.0            41        0   
...          ...        ...    ...           ...      ...   
14564  39.950060 -75.170129    2.0            10        0   
14565  39.968806 -75.180608    3.0             6        1   
14566  39.932756 -75.144504    3.0            32        1   
14567  39.945966 -75.169666    4.5            39        1   
14568  39.935982 -75.158665    4.5            35        1   

                                              attributes  \
0      {'RestaurantsDelivery': 'False', 'OutdoorSeati...   
1      {'RestaurantsReservations': 'True', 'Restauran...   
2      {'NoiseLevel': "u'quiet'", 'GoodForMeal': "{'d...   
3      {'Smoking': "u'no'", 'NoiseLevel': "u'average'...   
4      {'RestaurantsReservations': 'False', 'Caters':...   
...                                                  ...   
14564  {'BusinessParking': "{'garage': False, 'street...   
14565             {'BusinessAcceptsCreditCards': 'True'}   
14566  {'BusinessAcceptsCreditCards': 'True', 'Wheelc...   
14567  {'ByAppointmentOnly': 'True', 'GoodForKids': '...   
14568  {'WheelchairAccessible': 'False', 'Restaurants...   

                                              categories  \
0      Restaurants, Food, Bubble Tea, Coffee & Tea, B...   
1                      Sushi Bars, Restaurants, Japanese   
2                                    Korean, Restaurants   
3      Cocktail Bars, Bars, Italian, Nightlife, Resta...   
4                        Pizza, Restaurants, Salad, Soup   
...                                                  ...   
14564                Fashion, Shopping, Women's Clothing   
14565             Real Estate, Apartments, Home Services   
14566        Day Spas, Beauty & Spas, Skin Care, Massage   
14567  Health & Medical, Yoga, Shopping, Naturopathic...   
14568                               Restaurants, Mexican   

                                                   hours  
0      {'Monday': '7:0-20:0', 'Tuesday': '7:0-20:0', ...  
1      {'Tuesday': 

In [168]:
df1 = df1[['business_id', 'categories']]

In [169]:
df = pd.merge(df, df1, on='business_id', how='inner')

In [170]:
df.dropna(inplace=True)


In [171]:
df = df[df['categories'].str.contains('Restau|restau', case=False)]


keep only restaurant reviews

                            Step 1: label Reviews as positive or negative

In [175]:
df = df[['review_id', 'stars', 'text']]

If star is 5 or 4 its labeled positive and if its 1 or 2 its negative . Reviews that scoored 3 are getting dropped since they are in grey aerea

In [176]:
df['sentiment'] = df['stars'].apply(lambda x: 'positive' if x in [4, 5] else ('negative' if x in [1, 2] else None))
df = df.dropna(subset=['sentiment'])


/var/folders/gw/2h0k1fwd3x1cfqdc9_7xxxt80000gn/T/ipykernel_24330/528977032.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['sentiment'] = df['stars'].apply(lambda x: 'positive' if x in [4, 5] else ('negative' if x in [1, 2] else None))


In [177]:
df

review_id  stars  \
151     AqPFMleE6RsU23_auESxiA      5   
152     HME_ksGph3se7Aze5hxa-Q      2   
153     EJWyA5wpdVMji1j4TwSZqQ      5   
154     T_kAb2NeylB-JdNDKphryw      5   
155     NENaCqb6TNj5CyY1LOdI6Q      5   
...                        ...    ...   
967530  l6mqZx5do_w7dLBEeDWVDg      5   
967531  eYq5ckB294CqkRFqfR5gNg      5   
967532  RmmdxkfF9f77yLsMKvwTAA      5   
967533  S8lVRBNhs1XpTVQlahViYw      5   
967534  jH5Z2Z38c32uUogmPLxLVg      5   

                                                     text sentiment  
151     Wow!  Yummy, different,  delicious.   Our favo...  positive  
152     Dine-in gets 2 stars. Disappointing service & ...  negative  
153     After a long hiatus from reviewing I have awak...  positive  
154     We've eaten here 3 times and it seems that eac...  positive  
155     Came to Philly for a family event but stayed a...  positive  
...                                                   ...       ...  
967530  Very friendly people making great food at a re...  positive  
967531  I've had these pictures to post this review fo...  positive  
967532  Chicken stew is amazing, fresh juices are deli...  positive  
967533  Doesn't look like much when you pass by but th...  positive  
967534  Need to check out this place if you are in the...  positive  

[595601 rows x 4 columns]

a first approach is to use Logistic Regression like the step1 one in question 2 and classify its review as positive or negative

In [178]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import KFold
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import precision_recall_fscore_support
from sklearn.metrics import f1_score
from sklearn.metrics import confusion_matrix

In [179]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, f1_score, precision_recall_fscore_support
from sklearn.preprocessing import LabelEncoder
import pandas as pd
import numpy as np

le = LabelEncoder()
df['sentiment_label'] = le.fit_transform(df['sentiment'])

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(df['text'], df['sentiment_label'], test_size=0.2, random_state=42)

# Vectorize the text data using tf-idf
vectorizer = TfidfVectorizer(stop_words='english')
X_train_vectorized = vectorizer.fit_transform(X_train)
X_test_vectorized = vectorizer.transform(X_test)

# Fit a logistic regression model to the training data
logreg = LogisticRegression(random_state=0, solver='lbfgs', multi_class='auto', n_jobs=-1)
logreg.fit(X_train_vectorized, y_train)

# Make predictions on the test data
logreg_pred = logreg.predict(X_test_vectorized)

# Calculate evaluation metrics
logreg_precision, logreg_recall, logreg_fscore, logreg_support = precision_recall_fscore_support(y_test, logreg_pred, average='macro')
logreg_acc = logreg.score(X_test_vectorized, y_test)
logreg_f1score_score_last3 = f1_score(y_test, logreg_pred, average='macro', labels=[0, 1, 2])

# Print the evaluation metrics
print('Accuracy:', logreg_acc)
print('Precision:', logreg_precision)
print('Recall:', logreg_recall)
print('F1 Score:', logreg_fscore)

# Print the confusion matrix
cm_logreg = confusion_matrix(y_test, logreg_pred)
print('Confusion matrix:')
print(cm_logreg)

# Get the coefficients of the logistic regression model
feature_names = vectorizer.get_feature_names()
coefficients = logreg.coef_[0]
coef_df = pd.DataFrame({'feature': feature_names, 'coefficient': coefficients})
coef_df.sort_values(by='coefficient', ascending=False, inplace=True)
print("Words with highest positive weights:")
print(coef_df.head(20))
print("Words with lowest negative weights:")
print(coef_df.tail(20))


/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1580: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/util

Accuracy: 0.9569093610698366
Precision: 0.9440448757381743
Recall: 0.9239508996481278
F1 Score: 0.9335347864566745
Confusion matrix:
[[21673  3320]
 [ 1813 92315]]
Words with highest positive weights:
           feature  coefficient
33910    delicious    18.806718
7325       amazing    15.155920
52968        great    12.531357
88591      perfect    12.521350
42704    excellent    11.538361
11314      awesome    10.605073
14547         best    10.244786
44109    fantastic     9.721061
88608    perfectly     9.328183
44521     favorite     9.327861
70978        loved     8.864545
33553   definitely     8.581677
70968         love     8.413417
60409   incredible     8.304610
88600   perfection     7.660046
132989       yummy     7.438532
40973      enjoyed     7.214159
130800   wonderful     7.110894
52072         good     7.074575
89272   phenomenal     6.867040
Words with lowest negative weights:
               feature  coefficient
124080   underwhelming    -6.642343
83290              

In [181]:
positive_dictionary = []
negative_dictionary = []

for index, row in coef_df.iterrows():
    if row['coefficient'] > 1.:
        positive_dictionary.append(row['feature'])
    if row['coefficient'] < -2:
        negative_dictionary.append(row['feature'])

In [182]:
negative_dictionary.reverse()

Using  Log.Reg for classification gives us the coefficient that can be used as words for the dictionaries

Some other approach

LSA can extract expression so i am givving it a try

In [183]:
positive_reviews = df[df['sentiment'] == 'positive']
negative_reviews = df[df['sentiment'] == 'negative']

LSA can extract expression so i am givving it a try

In [184]:
positive_vectorizer = TfidfVectorizer(stop_words='english', ngram_range=(2, 2))
positive_X = positive_vectorizer.fit_transform(positive_reviews['text'])


In [185]:
from sklearn.decomposition import TruncatedSVD

positive_lsa = TruncatedSVD(n_components=3)
positive_X_lsa = positive_lsa.fit_transform(positive_X)

In [186]:
positive_feature_names = positive_vectorizer.get_feature_names()
positive_bigrams = []
for component in positive_lsa.components_:
    component_bigrams = []
    for index in component.argsort()[-10:]:
        component_bigrams.append(positive_feature_names[index])
    positive_bigrams.append(component_bigrams)


/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


In [188]:
positive_bigrams

[['love place',
  'really good',
  'food good',
  'good food',
  'happy hour',
  'highly recommend',
  'service great',
  'great service',
  'food great',
  'great food'],
 ['great prices',
  'great price',
  'great drinks',
  'place great',
  'atmosphere great',
  'great atmosphere',
  'service great',
  'great service',
  'food great',
  'great food'],
 ['good happy',
  'went happy',
  'hour great',
  'hour menu',
  'hour specials',
  'great service',
  'great happy',
  'food great',
  'great food',
  'happy hour']]

In [189]:
negative_vectorizer = TfidfVectorizer(stop_words='english', ngram_range=(2, 2))
negative_X = negative_vectorizer.fit_transform(negative_reviews['text'])

#  LSA on negative reviews
negative_lsa = TruncatedSVD(n_components=3)
negative_X_lsa = negative_lsa.fit_transform(negative_X)

# extract bigrams from negative LSA components
negative_feature_names = negative_vectorizer.get_feature_names()
negative_bigrams = []
for component in negative_lsa.components_:
    component_bigrams = []
    for index in component.argsort()[-10:]:
        component_bigrams.append(negative_feature_names[index])
    negative_bigrams.append(component_bigrams)

In [190]:
negative_bigrams

[['good food',
  'minutes later',
  'don know',
  '30 minutes',
  'tasted like',
  '10 minutes',
  '15 minutes',
  '20 minutes',
  'food good',
  'customer service'],
 ['great customer',
  'service horrible',
  'service skills',
  'good customer',
  'bad customer',
  'poor customer',
  'worst customer',
  'terrible customer',
  'horrible customer',
  'customer service'],
 ['cheese fries',
  'don waste',
  'better cheese',
  'don know',
  'customer service',
  'chicken cheese',
  'tasted like',
  'cheese steaks',
  'philly cheese',
  'cheese steak']]

I dont think these are reallt usable

Another approach is to use a common nlp to extract sentiment-bearing phrases from the text me and add them while using Textblot to positive and negative dictionaries 

In [194]:
import spacy
from textblob import TextBlob

In [195]:
nlp = spacy.load('en_core_web_sm')

In [196]:
positive_dict = {}
negative_dict = {}

In [225]:
counter=0

i have tried running it for all rows in the df but it took forever (produced over 11958494 phrases and i forced stoped it ) so i am gonna run a random sample of the df

In [250]:
df3 =df.sample(n=10000, random_state=1)

In [238]:
counter = 0
for index, row in df3.iterrows():
    doc = nlp(row['text'])
    counter = counter +1
    if counter in [1,2,3,4,5,6,10,100,1000,2000,5000,10000, 50000,55000]:
        print(counter)
    for chunk in doc.noun_chunks:
        phrases.append(chunk.text)
        if counter in [10,100,1000,2000,5000,10000, 50000]:
            print(counter)
    for token in doc:
        if counter in [10,100,1000,2000,5000,10000, 50000]:
            print(counter)
        if token.pos_ == 'ADJ':
            phrases.append(token.text)
#print to show where in the for loop is: from 5k to 10k it took 1,5 min

1
2
3
4
5
6
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
100
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000
1000

In [239]:
for phrase in phrases:
        blob = TextBlob(phrase)
        polarity = blob.sentiment.polarity
        if polarity > 0:
            if phrase in positive_dict:
                positive_dict[phrase] += 1
            else:
                positive_dict[phrase] = 1
        elif polarity < 0:
            if phrase in negative_dict:
                negative_dict[phrase] += 1
            else:
                negative_dict[phrase] = 1

In [245]:
negative_dictionary_blob = dict(sorted(negative_dict.items(), key=lambda item: item[1], reverse=True))

In [246]:
negative_dictionary_blob

{'other': 50146,
 'little': 47484,
 'small': 32989,
 'few': 30639,
 'bad': 20850,
 'long': 14165,
 'cold': 11258,
 'wrong': 9831,
 'hard': 9565,
 'least': 9381,
 'dry': 8722,
 'expensive': 7192,
 'chicken': 6704,
 'less': 6668,
 'green': 6336,
 'disappointed': 6298,
 'rude': 6163,
 'half': 5977,
 'late': 5949,
 'average': 5683,
 'cozy': 5671,
 'slow': 5526,
 'empty': 5520,
 'black': 5500,
 'bland': 5433,
 'terrible': 5206,
 'worst': 4999,
 'the chicken': 4944,
 'thin': 4406,
 'thick': 4302,
 'disappointing': 4290,
 'horrible': 4129,
 'single': 4025,
 'dark': 4018,
 'past': 3846,
 'casual': 3818,
 'sour': 3776,
 'heavy': 3732,
 'crazy': 3607,
 'typical': 3534,
 'mediocre': 3472,
 'poor': 3458,
 'usual': 3399,
 'weird': 3332,
 'raw': 2982,
 'limited': 2954,
 'awful': 2877,
 'tough': 2728,
 'sad': 2651,
 'Other': 2632,
 'difficult': 2603,
 'The chicken': 2521,
 'dirty': 2467,
 'plain': 2424,
 'sorry': 2395,
 'a long time': 2212,
 'flat': 2180,
 'negative': 2089,
 'worse': 2045,
 'ridiculo

In [247]:
positive_dictionary_blob = dict(sorted(positive_dict.items(), key=lambda item: item[1], reverse=True))

In [249]:
positive_dictionary_blob

{'good': 165135,
 'great': 122060,
 'delicious': 69936,
 'best': 54460,
 'nice': 53497,
 'more': 42147,
 'amazing': 41186,
 'friendly': 39580,
 'fresh': 35188,
 'first': 28924,
 'better': 27617,
 'sure': 26010,
 'Great': 25659,
 'perfect': 25016,
 'worth': 24158,
 'sweet': 24070,
 'happy': 23879,
 'many': 23719,
 'hot': 22997,
 'favorite': 21965,
 'excellent': 21571,
 'awesome': 19678,
 'new': 19449,
 'much': 17986,
 'special': 17720,
 'full': 17696,
 'large': 16786,
 'top': 15263,
 'huge': 14297,
 'fantastic': 13243,
 'attentive': 12998,
 'quick': 12990,
 'right': 12899,
 'wonderful': 12528,
 'busy': 12090,
 'free': 11958,
 'old': 11703,
 'high': 11658,
 'whole': 11436,
 'most': 10844,
 'cool': 10549,
 'decent': 10544,
 'able': 10402,
 'warm': 10345,
 'own': 10178,
 'clean': 9156,
 'Good': 9054,
 'authentic': 8504,
 'real': 8258,
 'soft': 8078,
 'fine': 7945,
 'reasonable': 7627,
 'cheap': 7571,
 'beautiful': 7464,
 'main': 7173,
 'unique': 7060,
 'ready': 7059,
 'incredible': 6804,
 

In [251]:
def classify_review(text):
    doc = nlp(text)
    # get noun chunks and adjectival phrases
    phrases = []
    for chunk in doc.noun_chunks:
        phrases.append(chunk.text)
    for token in doc:
        if token.pos_ == 'ADJ':
            phrases.append(token.text)

    # count the number of positive and negative words in the text
    num_positive = 0
    num_negative = 0
    for phrase in phrases:
        if phrase in positive_dict:
            num_positive += positive_dict[phrase]
        if phrase in negative_dict:
            num_negative += negative_dict[phrase]

    # classify the review based on the number of positive and negative words
    if num_positive > num_negative:
        return 'Positive'
    elif num_positive < num_negative:
        return 'Negative'
    else:
        return None


A good way to evaluate if the dictionaries is to use them to classify some reviews that we already know the sentiment.

In [264]:
#avoiding overfitting 
exclude_ids = df3['review_id'].tolist()
df_test=df[~df['review_id'].isin(exclude_ids)].sample(n=30000, random_state=1)

In [265]:
df_test['sentiment_label_pred'] = df_test['text'].apply(classify_review)

In [ ]:
#RMSE calculation
actual_ratings = df_test['stars'].values
predicted_ratings = df_test['ua_predicted'].values

# Calculate the RMSE
rmse = np.sqrt(mean_squared_error(actual_ratings, predicted_ratings))
print("RMSE:", rmse)

In [273]:
df_test['sentiment_label_pred']= df_test['sentiment_label_pred'].str.lower()

In [280]:
from sklearn.metrics import accuracy_score
df_test_1 = df_test.copy()
df_test_1.loc[df_test_1['sentiment_label_pred'].isnull() & (df_test_1['sentiment_label'] == 'positive'), 'sentiment_label_pred'] = 'negative'
df_test_1.loc[df_test_1['sentiment_label_pred'].isnull() & (df_test_1['sentiment_label'] == 'negative'), 'sentiment_label_pred'] = 'positive'
df_test_1['sentiment_label_pred'] = le.fit_transform(df_test_1['sentiment_label_pred'])
accuracy = accuracy_score(df_test_1['sentiment_label'], df_test_1['sentiment_label_pred'])
print("Accuracy:", accuracy)

Accuracy: 0.8035333333333333


An accuracy score of 0.8035333333333333 is a good result and suggests that this model is likely to be useful for making predictions on new data.
Increasing the size of the sample could potentially improve the accuracy of the dictionary, as there would be  more data points to draw from. However, it's important to keep in mind that the accuracy of a dictionary depends on various factors, such as the quality of the data collected, the relevance of the sources used, and the criteria used to classify and categorize the data. It's important to consider the ethical implications of using a dictionary that may produce results harmful to certain marginalized groups. It's important to approach the creation of any tool or resource with sensitivity and awareness of potential biases or harmful effects it may have on certain populations. Similarly, the location-specific nature of the data could impact the accuracy of the dictionary when applied to other locations.Lastly, using a subsidized sample may limit the ability of the dictionary to recognize certain slang terms or variations in language use.